# Data Consistency Checks 

In [1]:
# Import libraries 
import pandas as pd 
import numpy as np 
import os

In [2]:
path = r'/Users/nurgul/Instacart Basket Analysis'

In [3]:
# Imported Data set products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col=False)

In [4]:
# Imported Data set orders_wrangled.csv
df_ords=pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col=False)

In [5]:
# Create a dataframe
df_test=pd.DataFrame()

In [6]:
# Create a mixed type column
df_test['mix']=['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis=1)
    if len (df_test[weird])>0:
        print(col)

mix


/var/folders/b1/dcgm49cs1dx0p3lkf95fklvr0000gp/T/ipykernel_95556/222267165.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis=1)


In [9]:
df_test['mix']=df_test['mix'].astype('str')

In [10]:
#Finding Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
#Addressing missing values
df_nan=df_prods[df_prods['product_name'].isnull()==True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
df_prods.shape

(49693, 5)

In [14]:
df_prods_clean=df_prods[df_prods['product_name'].isnull()==False]

In [15]:
df_prods_clean.shape

(49677, 5)

In [16]:
#Finding full duplicates
df_dups=df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
#Addressing full duplicates
#checking current number of rows
df_prods_clean.shape

(49677, 5)

In [19]:
df_prods_clean_no_dups=df_prods_clean.drop_duplicates()

In [20]:
df_prods_clean_no_dups.shape

(49672, 5)

In [21]:
df_prods_checked=df_prods_clean_no_dups

In [22]:
df_prods_checked.to_csv(os.path.join(path,'02 Data','Prepared Data', 'products_checked.csv'))

# Exercise 4.5 Data Consistency Checks¶

In [23]:
# Q.2 Run the df.describe() function on your df_ords dataframe
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  hour_of_day_ordered  days_since_prior_order  
count        3.421083e+06         3.421083e+06            3.214874e+06  
mean         2.776219e+00         1.345202e+01            1.111484e+01  
std          2.046829e+00         4.226088e+00            9.206737e+00  
min          0.000000e+00         0.000000e+00            0.000000e+00  
25%          1.000000e+00         1.000000e+01            4.000000e+00  
50%          3.000000e+00         1.300000e+01            7.000000e+00  
75%          5.000000e+00         1.600000e+01            1.500000e+01  
max          6.000000e+00         2.300000e+01            3.000000e+01

In [24]:
df_ords

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
1                 1   2398795        1    prior             2   
2                 2    473747        1    prior             3   
3                 3   2254736        1    prior             4   
4                 4    431534        1    prior             5   
...             ...       ...      ...      ...           ...   
3421078     3421078   2266710   206209    prior            10   
3421079     3421079   1854736   206209    prior            11   
3421080     3421080    626363   206209    prior            12   
3421081     3421081   2977660   206209    prior            13   
3421082     3421082    272231   206209    train            14   

         orders_day_of_week  hour_of_day_ordered  days_since_prior_order  
0                         2                    8                     NaN  
1                         3                    7                    15.0  
2                         3                   12                    21.0  
3                         4                    7                    29.0  
4                         4                   15                    28.0  
...                     ...                  ...                     ...  
3421078                   5                   18                    29.0  
3421079                   4                   10                    30.0  
3421080                   1                   12                    18.0  
3421081                   1                   12                     7.0  
3421082                   6                   14                    30.0  

[3421083 rows x 8 columns]

Looking at the min and max of "days since prior order," I've noticed inconsistent values. The min value is 0, and the max is 3. However, 25 and 50 percentiles are higher than the max value (4 and 7). 
Same with "order number," the min and max value is 1, though 25 percentile is 5, which is higher than the max.  


In [25]:
# Q.3 Check for mixed-type data in your df_ords dataframe.
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

/var/folders/b1/dcgm49cs1dx0p3lkf95fklvr0000gp/T/ipykernel_95556/3793779795.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/b1/dcgm49cs1dx0p3lkf95fklvr0000gp/T/ipykernel_95556/3793779795.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/b1/dcgm49cs1dx0p3lkf95fklvr0000gp/T/ipykernel_95556/3793779795.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
/var/folders/b1/dcgm49cs1dx0p3lkf95fklvr0000gp/T/ipykernel_95556/3793779795.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].ilo

There are no mixed data type in df_ords data frame

In [26]:
# Q.5 Run a check for missing values in your df_ords dataframe.
#Finding Missing Values
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
hour_of_day_ordered            0
days_since_prior_order    206209
dtype: int64

In [27]:
df_ords_nan=df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [28]:
df_ords_nan

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  hour_of_day_ordered  days_since_prior_order  
0                         2                    8                     NaN  
11                        2                   11                     NaN  
26                        1                   14                     NaN  
39                        6                   11                     NaN  
45                        3                   12                     NaN  
...                     ...                  ...                     ...  
3420930                   4                   12                     NaN  
3420934                   3                   18                     NaN  
3421002                   6                   19                     NaN  
3421019                   1                   15                     NaN  
3421069                   3                   11                     NaN  

[206209 rows x 8 columns]

According to the data frame, there are 206209 missing values in "days_since_prior_order." The number is the same as the user ID numbers. This implies that these users have ordered for the first time, and each will have a missing value for the days since the prior order. There is no need to address missing data in this context;  this will help to see users who have ordered for the first time and the frequency of their orders after the first time. 

In [29]:
#Q.7 Run a check for duplicate values in your df_ords data
#Finding full duplicates
df_ords_dups=df_ords[df_ords.duplicated()]

In [30]:
# Viewing duplicates values for df_ords
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, orders_day_of_week, hour_of_day_ordered, days_since_prior_order]
Index: []

No duplicates found

### Q.9 Export Data frame df_ords and df_prods

In [31]:
# Export df_ords as orders_checked
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [32]:
# Export df_prods as products_checked
df_prods_checked.to_csv(os.path.join(path,'02 Data','Prepared Data', 'products_checked.csv'))